include/xtensor/xbuilder.hpp

```diff
@@ -600,61 +633,28 @@ namespace xt
             }
 
             template <class... Args>
-            inline value_type operator()(std::size_t first_idx, Args... args) const
+            inline value_type operator()(Args... args) const
             {
-                return m_source(m_shape_first - first_idx, args...);
+                return access_impl(xindex({ args... }));
             }
 
             template <class It>
             inline value_type element(It first, It last) const
             {
                 // TODO: avoid memory allocation
-                xindex idx(first, last);
-                idx.front() = m_shape_first - idx.front();
-                return m_source.element(idx.begin(), idx.end());
+                return access_impl(xindex(first, last));
             }
 
         private:
-            CT m_source;
-            const size_type m_shape_first;
-        };
-
-        template <class CT>
-        struct fliplr_impl
-        {
-            using xexpression_type = std::decay_t<CT>;
-            using value_type = typename xexpression_type::value_type;
-            using size_type = typename xexpression_type::size_type;
-
-            fliplr_impl(CT source)
-                : m_source(source), m_shape_last(m_source.shape().back() - 1)
-            {
-            }
-
-            inline value_type operator()() const
-            {
-                // special case when called without args (happens when printing)
-                return value_type();
-            }
-
-            template <class... Args>
-            inline value_type operator()(Args... args, std::size_t last_idx) const
-            {
-                return m_source(args..., m_shape_last - last_idx);
-            }
-
-            template <class It>
-            inline value_type element(It first, It last) const
+            inline value_type access_impl(xindex idx) const
             {
-                // TODO: avoid memory allocation
-                xindex idx(first, last);
-                idx.back() = m_shape_last - idx.back();
+                idx[m_axis] = m_shape_at_axis - idx[m_axis];
                 return m_source.element(idx.begin(), idx.end());
             }
```

Comment:
Here no since you return a value and not a reference. To be rigorous we should call ``element`` with ``const_iterator`` built upon ``first`` and ``last``.